File: python/roman-numerals/roman_numerals.py
'''
Roman numerals code kata.
'''

def to_roman(number):
  """
  Converts an arabic number into a roman number.

  @type  number: integer.
  @param number: The arabic number to convert.

  @rtype  string
  @return The roman number.
  """
  return _to_roman(number, "")

_ALPHABET = [(1000, "M"), (900, "CM"), (500, "D"), (400, "CD"),
             ( 100, "C"), ( 90, "XC"), ( 50, "L"), ( 40, "XL"),
             ( 10,  "X"), (  9, "IX"), (  5, "V"), (  4, "IV"),
             (  1,  "I"), (  0,  "")]

def _to_roman(number, result):
  if number == 0:
    return result
  else:
    digit_value, digit_symbol = next((v, s) for (v, s) in _ALPHABET if number >= v)
    return _to_roman(number - digit_value, result + digit_symbol)

File: ruby/roman-numerals/roman_numerals.rb
#
# Roman numerals code kata.
#

ROMAN_ALPHABET = {1000 => "M", 900 => "CM", 500 => "D", 400 => "CD",
                   100 => "C",  90 => "XC",  50 => "L",  40 => "XL",
                    10 => "X",   9 => "IX",   5 => "V",   4 => "IV",
                     1 => "I",   0 =>  "" }

# Converts a arabic number in it's roman string representation.
#
# @param number [Fixnum] The arabic number to convert.
#
# @return [String] The roman number as a string.
def to_roman(number)
  result = ""

  while number > 0 do
    value, symbol = ROMAN_ALPHABET.find { |value, symbol| number >= value }

    result += symbol
    number -= value
  end

  return result
end
